File: src/main/java/com/tkr/ledger/CheckpointLedger.java
package com.tkr.ledger;

import com.tkr.types.WireTypes.*;
import com.tkr.types.WireTypes.Status;
import com.tkr.util.DigestUtil;

import java.util.ArrayList;
import java.util.List;

/**
 * Checkpoint ledger for session merge digest sealing and chain verification.
 * Ported from legacy-cpp checkpoint_ledger.cc.
 */
public final class CheckpointLedger {

    public static final class CheckpointRecord {
        public int checkpointSeq;
        public int sessionId;
        public int legCount;
        public int mergeDigest;
        public int prevCheckpointSeq;
        public long sealedAtNs;
        public boolean valid;
    }

    public static final class CheckpointLedgerConfig {
        public int maxCheckpoints = 1024;
        public boolean verifyChain = true;
    }

    public static final class CheckpointSealResult {
        public Status status = Status.OK;
        public int checkpointSeq;
        public int mergeDigest;
        public boolean chainValid;
    }

    public static final class CheckpointVerifyResult {
        public Status status = Status.OK;
        public int checkpointsVerified;
        public boolean chainIntact;
        public int firstGapSeq;
    }

    private static final CheckpointLedger GLOBAL = new CheckpointLedger(new CheckpointLedgerConfig());

    public static CheckpointLedger global() {
        return GLOBAL;
    }

    private final CheckpointLedgerConfig config;
    private final List<CheckpointRecord> records = new ArrayList<>();
    private int latestSeq;

    public CheckpointLedger() {
        this(new CheckpointLedgerConfig());
    }

    public CheckpointLedger(CheckpointLedgerConfig config) {
        this.config = config != null ? config : new CheckpointLedgerConfig();
    }

    public int computeMergeDigest(SessionWireFrame frame) {
        int digest = DigestUtil.FNV_OFFSET;
        digest = DigestUtil.mixDigest(digest, frame.header.sessionId);
        digest = DigestUtil.mixDigest(digest, frame.header.legCount);
        digest = DigestUtil.mixDigest(digest, frame.header.checkpointSeq);
        digest = DigestUtil.mixDigest(digest, frame.header.flags);
        for (WireSessionLeg leg : frame.legs) {
            digest = DigestUtil.mixDigest(digest, leg.legId);
            digest = DigestUtil.mixDigest(digest, leg.clOrdId);
            digest = DigestUtil.mixDigest(digest, leg.allocAccount);
            digest = DigestUtil.mixDigest(digest, leg.qtyMilli);
            digest = DigestUtil.mixDigest(digest, leg.flags);
            if (leg.refLen > 0 && leg.refOffset < frame.refBlob.length) {
                int avail = frame.refBlob.length - leg.refOffset;
                int len = Math.min(leg.refLen, avail);
                digest = DigestUtil.mixDigest(digest, DigestUtil.fnv1a32(frame.refBlob, leg.refOffset, len));
            }
        }
        for (MergeSlot slot : frame.mergeSlots) {
            if (!slot.pinned) continue;
            digest = DigestUtil.mixDigest(digest, slot.slotId);
            digest = DigestUtil.mixDigest(digest, slot.legId);
            digest = DigestUtil.mixDigest(digest, slot.checkpointSeq);
            if (slot.refData != null && slot.refLen > 0) {
                digest = DigestUtil.mixDigest(digest, DigestUtil.fnv1a32(slot.refData, 0, slot.refLen));
            }
        }
        return digest;
    }

    private void evictOldestIfNeeded() {
        if (records.size() >= config.maxCheckpoints) records.remove(0);
    }

    private boolean validateSequenceChain() {
        if (records.isEmpty()) return true;
        int expectedPrev = 0;
        for (CheckpointRecord rec : records) {
            if (rec.checkpointSeq != expectedPrev + 1) return false;
            if (rec.prevCheckpointSeq != expectedPrev) return false;
            if (!rec.valid) return false;
            expectedPrev = rec.checkpointSeq;
        }
        return true;
    }

    public CheckpointSealResult sealSession(SessionWireFrame frame) {
        CheckpointSealResult result = new CheckpointSealResult();
        evictOldestIfNeeded();
        CheckpointRecord record = new CheckpointRecord();
        record.checkpointSeq = latestSeq + 1;
        record.sessionId = frame.header.sessionId;
        record.legCount = frame.header.legCount;
        record.mergeDigest = computeMergeDigest(frame);
        record.prevCheckpointSeq = latestSeq;
        record.sealedAtNs = (long) record.checkpointSeq * 1_000_000L;
        record.valid = true;
        if (config.verifyChain && latestSeq > 0) {
            CheckpointRecord prev = lookup(latestSeq);
            if (prev.sessionId != 0 && prev.sessionId != record.sessionId) record.valid = false;
        }
        records.add(record);
        latestSeq = record.checkpointSeq;
        result.checkpointSeq = record.checkpointSeq;
        result.mergeDigest = record.mergeDigest;
        result.chainValid = validateSequenceChain();
        result.status = result.chainValid ? Status.OK : Status.SESSION_GAP;
        return result;
    }

    public CheckpointVerifyResult verifyChain() {
        CheckpointVerifyResult result = new CheckpointVerifyResult();
        result.checkpointsVerified = records.size();
        result.chainIntact = validateSequenceChain();
        result.status = result.chainIntact ? Status.OK : Status.SESSION_GAP;
        if (!result.chainIntact) {
            int expected = 0;
            for (CheckpointRecord rec : records) {
                if (rec.checkpointSeq != expected + 1) {
                    result.firstGapSeq = expected + 1;
                    break;
                }
                expected = rec.checkpointSeq;
            }
        }
        return result;
    }

    public CheckpointRecord lookup(int checkpointSeq) {
        for (CheckpointRecord rec : records) {
            if (rec.checkpointSeq == checkpointSeq) return rec;
        }
        return new CheckpointRecord();
    }

    public void reset() {
        records.clear();
        latestSeq = 0;
    }

    public int latestCheckpointSeq() {
        return latestSeq;
    }

    public List<CheckpointRecord> snapshot() {
        return new ArrayList<>(records);
    }

    public boolean hasGapAfter(int seq) {
        for (CheckpointRecord rec : records) {
            if (rec.prevCheckpointSeq == seq && rec.checkpointSeq != seq + 1) return true;
        }
        return false;
    }

    public int digestAt(int checkpointSeq) {
        CheckpointRecord rec = lookup(checkpointSeq);
        return rec.mergeDigest;
    }

    public Status pinCheckpoint(int checkpointSeq, SessionWireFrame frame) {
        CheckpointRecord rec = lookup(checkpointSeq);
        if (rec.checkpointSeq == 0) return Status.BOUNDS_ERROR;
        int computed = computeMergeDigest(frame);
        return computed == rec.mergeDigest ? Status.OK : Status.BOUNDS_ERROR;
    }

    public String summarizeChain() {
        return "checkpoints=" + records.size() + " latest=" + latestSeq + " valid=" + validateSequenceChain();
    }
}

File: legacy-cpp/src/ledger/checkpoint_ledger.cc
#include "tkr/ledger/checkpoint_ledger.h"

#include "tkr/util/bounds.h"

#include <algorithm>

namespace tkr {
namespace ledger {
namespace {

constexpr std::uint32_t kFnvOffset = 2166136261u;
constexpr std::uint32_t kFnvPrime = 16777619u;

std::uint32_t MixField(std::uint32_t hash, std::uint32_t value) {
  hash ^= value;
  hash *= kFnvPrime;
  return hash;
}

}  // namespace

CheckpointLedger& GlobalCheckpointLedger() {
  static CheckpointLedger ledger(CheckpointLedgerConfig{1024, true});
  return ledger;
}

CheckpointLedger::CheckpointLedger(CheckpointLedgerConfig config)
    : config_(config), latest_seq_(0) {}

std::uint32_t CheckpointLedger::ComputeMergeDigest(
    const SessionWireFrame& frame) const {
  std::uint32_t digest = kFnvOffset;

  digest = MixField(digest, frame.header.session_id);
  digest = MixField(digest, frame.header.leg_count);
  digest = MixField(digest, frame.header.checkpoint_seq);
  digest = MixField(digest, frame.header.flags);

  for (const WireSessionLeg& leg : frame.legs) {
    digest = MixField(digest, leg.leg_id);
    digest = MixField(digest, leg.cl_ord_id);
    digest = MixField(digest, leg.alloc_account);
    digest = MixField(digest, leg.qty_milli);
    digest = MixField(digest, leg.flags);

    if (leg.ref_len > 0 && leg.ref_offset < frame.ref_blob.size()) {
      const std::size_t avail = frame.ref_blob.size() - leg.ref_offset;
      const std::size_t len =
          (leg.ref_len <= avail) ? leg.ref_len : avail;
      digest = MixField(digest,
                        util::Fnv1a32(frame.ref_blob.data() + leg.ref_offset,
                                      len));
    }
  }

  for (const MergeSlot& slot : frame.merge_slots) {
    if (!slot.pinned) {
      continue;
    }
    digest = MixField(digest, slot.slot_id);
    digest = MixField(digest, slot.leg_id);
    digest = MixField(digest, slot.checkpoint_seq);
    if (slot.ref_ptr != nullptr && slot.ref_len > 0) {
      digest = MixField(digest, util::Fnv1a32(slot.ref_ptr, slot.ref_len));
    }
  }

  return digest;
}

void CheckpointLedger::EvictOldestIfNeeded() {
  if (records_.size() < config_.max_checkpoints) {
    return;
  }
  records_.erase(records_.begin());
}

bool CheckpointLedger::ValidateSequenceChain() const {
  if (records_.empty()) {
    return true;
  }

  std::uint32_t expected_prev = 0;
  for (const CheckpointRecord& rec : records_) {
    if (rec.checkpoint_seq != expected_prev + 1) {
      return false;
    }
    if (rec.prev_checkpoint_seq != expected_prev) {
      return false;
    }
    if (!rec.valid) {
      return false;
    }
    expected_prev = rec.checkpoint_seq;
  }
  return true;
}

CheckpointSealResult CheckpointLedger::SealSession(
    const SessionWireFrame& frame) {
  CheckpointSealResult result{};
  EvictOldestIfNeeded();

  CheckpointRecord record{};
  record.checkpoint_seq = latest_seq_ + 1;
  record.session_id = frame.header.session_id;
  record.leg_count = frame.header.leg_count;
  record.merge_digest = ComputeMergeDigest(frame);
  record.prev_checkpoint_seq = latest_seq_;
  record.sealed_at_ns =
      static_cast<std::uint64_t>(record.checkpoint_seq) * 1000000u;
  record.valid = true;

  if (config_.verify_chain && latest_seq_ > 0) {
    const CheckpointRecord prev = Lookup(latest_seq_);
    if (prev.session_id != 0 && prev.session_id != record.session_id) {
      record.valid = false;
    }
  }

  records_.push_back(record);
  latest_seq_ = record.checkpoint_seq;

  result.checkpoint_seq = record.checkpoint_seq;
  result.merge_digest = record.merge_digest;
  result.chain_valid = ValidateSequenceChain();
  result.status = result.chain_valid ? Status::kOk : Status::kSessionGap;
  return result;
}

CheckpointVerifyResult CheckpointLedger::VerifyChain() const {
  CheckpointVerifyResult result{};
  result.checkpoints_verified =
      static_cast<std::uint32_t>(records_.size());
  result.chain_intact = ValidateSequenceChain();
  result.status = result.chain_intact ? Status::kOk : Status::kSessionGap;

  if (!result.chain_intact) {
    std::uint32_t expected = 0;
    for (const CheckpointRecord& rec : records_) {
      if (rec.checkpoint_seq != expected + 1) {
        result.first_gap_seq = expected + 1;
        break;
      }
      expected = rec.checkpoint_seq;
    }
  }

  return result;
}

CheckpointRecord CheckpointLedger::Lookup(std::uint32_t checkpoint_seq) const {
  for (const CheckpointRecord& rec : records_) {
    if (rec.checkpoint_seq == checkpoint_seq) {
      return rec;
    }
  }
  return CheckpointRecord{};
}

void CheckpointLedger::Reset() {
  records_.clear();
  latest_seq_ = 0;
}

}  // namespace ledger
}  // namespace tkr
